mlfinlab/feature_importance/fingerpint.py

```diff
@@ -0,0 +1,181 @@
+    """
+    Regression Fingerprint class
+    """
+
+    def __init__(self, model: object, X: pd.DataFrame, num_values=50):
+        """
+        Constructs Regression model fingerprint.
+        :param model: (object) trained regression model
+        :param X: (pd.DataFrame) of features
+        :param num_values: (int) number of values to fix for each feature
+        """
+
+        self.X = X
+        self.model = model
+
+        # Effects containers
+        self.linear_effect = None
+        self.non_linear_effect = None
+        self.pair_wise_effect = None
+
+        self.num_values = num_values
+
+        # TODO: drop binary features?
+        # Get possible feature values (Step 1)
+        self.feature_values = {}  # Dictionary of feature values range used to build dependence functions
+        for feature in self.X.columns:
+            values = []
+            for q in np.linspace(0, 1, self.num_values):
+                values.append(np.quantile(self.X[feature], q=q))
+            self.feature_values[feature] = np.array(values)
+
+        # DataFrame with partial dependence function values for columns from X (f_k)
+        self.ind_partial_dep_functions = pd.DataFrame(index=list(range(self.num_values)), columns=self.X.columns)
+
+        self.feature_column_position_mapping = dict(zip(self.X.columns, range(0, self.X.shape[1])))
+
+        # Get partial dependency functions
+        self._get_individual_partial_dependence()
+
+    def _get_individual_partial_dependence(self) -> None:
+        """
+        Get individual partial dependence function values for each column.
+        :return: None
+        """
+        for col in self.X.columns:
+            y_mean_arr = []  # Array of mean(prediction) for each feature value
+            for x_k in self.feature_values[col]:
+                # Step 2
+                col_k_position = self.feature_column_position_mapping[col]
+                X_ = self.X.values.copy()
+                X_[:, col_k_position] = x_k
+
+                # Step 3
+                y_pred = self.model.predict(X_)
+
+                y_pred_mean = y_pred.mean()
+
+                y_mean_arr.append(y_pred_mean)
+
+            self.ind_partial_dep_functions[col] = y_mean_arr
+
+    def _get_linear_effect_estimation(self) -> dict:
+        """
+        Get linear effect estimates.
+        :return: (dict) of linear effect estimates for each feature column.
+        """
+        store = {}
+        for col in self.X.columns:
+            x = self.feature_values[col].reshape(-1, 1)
+            y = self.ind_partial_dep_functions[col]
+
+            lmodel = LinearRegression(fit_intercept=True, normalize=False)
+            lmodel.fit(x, y)
+            y_mean = np.mean(y)
+            linear_effect = np.mean(np.abs(lmodel.predict(x) - y_mean))
+            store[col] = linear_effect
+        return store
+
+    def _get_non_linear_effect_estimation(self) -> dict:
+        """
+        Get non-linear effect estimates.
```

Comment:
Perhaps we can will out these docstrings by including snippets from the paper that will explain it in depth. Thus the code becomes an extension of the paper.

Reply:
You mean to add formulas?

Reply:
Just to provide an elaborate explanation from the paper on what each effect captures and the idea behind it.